### PR TITLE
Install Python bindings with pip 

### LINKED ISSUE #1812: Workaround for making RDKit can be installed by pip directly
As title, I am trying to make `RDKit` installable by `pip` directly, which is the recommend installation for a python package. To do this, we need to add a `setup.py`, a `MANIFEST.in`, and probably need to put more settings into `setup.cfg`.

The main challenge of this is that `RDKit` is natively built by `CMake`, which is not directly supported by Python standard library. Therefore, we need to extend the build extension mechanism for `CMake`. There are two projects I found that use `CMake` as well: [cmake_example](https://github.com/pybind/cmake_example) and [boost-python-cmake-mac](https://github.com/buddha314/boost-python-cmake-mac). There are also other resources like [stackoverflow question](https://stackoverflow.com/questions/42585210/extending-setuptools-extension-to-use-cmake-in-setup-py) and [quora question](https://www.quora.com/How-do-you-distribute-Boost.Python-modules).

I put my current work here: https://github.com/hsiaoyi0504/rdkit/tree/d442394c45fb05d4f989c7f031ab67213ad1536d. However, I failed to finish. In principle, that is not so hard to extend the build extension process. We just need to call some CMake commands and put the result in `build_temp` directory. One observation I found is that after running
``` shell
mkdir build
cd build
cmake ..
cmake --build  # or make 
```

In [cmake_example](https://github.com/pybind/cmake_example), it will produce a file `cmake_example.cpython-36m-darwin.so`, but for our RDkit, it doesn't have a file like that.
@greglandrum Do you know how to produce such a file ? Where should I edit the `CMakeLists.txt`?

- RDKit Version: based on `Release_2017_09_3`.
- Platform: MacOS, but should be the same on Linux.

### PR DESCRIPTION
**Pull request by maksbotan**
#### Reference Issue

Fixes #1812 (in a way).

#### What does this implement/fix? Explain your changes.

Currently RDKit installs its Python bindings simply by copying .py files
and C++ extensions to appropriate site-packages directory.

This approach has a disadvantage -- such installations are not
registered in Python's package manager pip and thus may other packages
may not specify 'rdkit' as their dependency. And it is often desirable
to specify all your dependencies in setup.py or pyproject.toml file.

This commit fixes this by packaging all Python modules in a .whl package
and then installing it with pip.

Unfortunately, there is no way to make this .whl self-contained by
statically linking Boost::Python and RDKit itself. However, having a
separate wheel file is still an advantage, since one may have one global
install of RDKit (for example, with brew) and then install the .whl in
virtual environments needing RDKit.

#### Any other comments?

This is needed for our company to integrate RDKit in our CI/CD processes more easily.

I've also removed `Code/GraphMol/Descriptors/test_data/getaway.out` file, since

1. There is `GETAWAY.out` in the same directory and it messes up on case-insensitive systems like macOS.
2. The two files are identical.
3. Neither is used anyway.

I'm also going to port formula in https://github.com/rdkit/homebrew-rdkit to install RDKit this way.

### CODE REVIEW

**Comment by greglandrum** [comment]
Why is it necessary to move the `rdkit` directory into `python/rdkit`?

**Comment by maksbotan** [comment]
I moved `rdkit` to `python` directory to make this directory structure for setuptools:

```
setup.py
rdkit/
  __init__.py
  Chem/
  ...
```

Theoretically, we can do this only in `${CMAKE_BINARY_DIR}`, avoiding the move in source dir, but this will break in-source (without `build` directory) builds. If this is OK for you, I'll move back.

**Comment by maksbotan** [comment]
Great, the tests have passed. I will fix Java tests to refer to the new location of test data. And I'll also add Travis job for `INTREE=OFF` variant to make sure my new code works as expected.

**Comment by maksbotan** [comment]
@greglandrum looks like CI is good :)

**Comment by bp-kelley** [comment]
I’m don’t think we need to move rdkit into /python, this is going to break a lot of builds that run the regression tests. (Conda/brew perhaps)

Can’t setup.py just set the appropriate search directory?

**Comment by maksbotan** [comment]
So, it seems to work with original layout.

**Comment by maksbotan** [comment]
I've updated brew formula for rdkit for my changes at https://github.com/rdkit/homebrew-rdkit/pull/76.
conda-forge package should not require any updated, but I'll check that later just in case.

What are you thoughts on this now, @greglandrum @bp-kelley ?

**Comment by UnixJunkie** [comment]
Is this related to creating a pip installer for rdkit?
People have expressed interest for it in the past.

**Comment by maksbotan** [comment]
@UnixJunkie this only enables pip to "see" that rdkit is installed.

Sadly, making self-contained rdkit wheel is rather challenging. As far as I can tell, you cannot link boost-python statically. I tried and I failed - I achieved linking without warnings, but it all broke in runtime. 
Another way is to bundle boost dynamic libraries in the wheel. This most certainly can be done. If you like this idea, I can do that. But it doesn't seem too nice for me.

**Comment by bp-kelley** [comment]
I've actually successfully made a wheel with the binaries.  I was hoping to get this out by the RDKit user group meeting.

Do you want to play with this new wheel to see if it is a better solution? 

 If it ends up not working, I have some reservations with this commit since it radically changes the build structure and forces changes in downstream builds like conda/brew but I think we can resolve them.

**Comment by maksbotan** [comment]
I checked it in conda-forge [package](https://github.com/conda-forge/rdkit-feedstock/tree/master/recipe), changing URL to my branch and building the package, and it works:

```console
$ conda list | grep rdkit
# packages in environment at /Users/maksbotan/miniconda3/envs/rdkit-local:
rdkit                     2019.03.4        py37h9537418_1    local
$ pip freeze | grep rdkit
rdkit==2019.9.1.dev1
$ python3 -c 'import rdkit; print(rdkit)'
<module 'rdkit' from '/Users/maksbotan/miniconda3/envs/rdkit-local/lib/python3.7/site-packages/rdkit/__init__.py'>
```

(conda-reported version is wrong as I did not change it in the recipe).

Are there any more news on this?

**Comment by maksbotan** [comment]
So, are there any news on this?

**Comment by bp-kelley** [comment]
I have some minor progress to report tomorrow, in that I have some pip installable Ubuntu packages.  As long as we document that this change here will force third party builds to be updated.  I know a few companies with internal builds that will need to update,  but the changes aren’t hard.

**Comment by greglandrum** [comment]
I am going to be very, very reluctant to agree to merge anything that results in changes having to be made to other peoples' build processes, but I will reserve judgement until I see what you guys come up with.

**Comment by maksbotan** [comment]
@greglandrum well, that is inevitable consequence of innovation and improvement :)
I, however, tried to make the transition as seamless as possible -- if one uses `make install` to install RDKit, it will still work, since my cmake scripts just call `pip install`.

One thing that gets broken by this is building of deb packages and their like. However, I have an idea how to fix that as well, I'll push it here in the next couple hours.

**Comment by UnixJunkie** [comment]
Oh yes, please don't break the building of deb packages.
On Linux, you can use this script: Scripts/create_deb_packages.sh
to check they can still be built.
Also, you need to install the produced debs and test rdkit can be loaded in a python interpreter
as a final check.

**Comment by greglandrum** [comment]
> @greglandrum well, that is inevitable consequence of innovation and improvement :)

I'm going to be diplomatic and not share what I think about pip here. :-)

**Comment by maksbotan** [comment]
Okay, take a look at the latest commit. With it `create_deb_packages.sh` produces the expected result:

```console
# dpkg -c RDKit-2019.09.1dev1-Linux-Python.deb | grep dist-info
drwxr-xr-x root/root         0 2019-10-30 15:30 ./usr/lib/python3.6/site-packages/rdkit-2019.9.1.dev1.dist-info/
```

As far as I can tell from my experience with package managers, this fix should be enough for most of them. Even my earlier homebrew PR is now unneccessary.

And yes, of course I agree that `pip` is a horrible piece of __software__, but that's all what we have in Python world and that's what we have to use, so it would be nice to support it. At least for our company this will be a considerable improvement.

**Comment by kuparez** [comment]
Is there any update to this?)

It would have been really nice to have this approach as an alternative to the Conda.

**Comment by njzjz** [comment]
Will this `whl` file available in https://pypi.org?

**Comment by greglandrum** [comment]
I just spent a couple of hours looking at this, getting it to build, and trying it out and I'm curious what problem it's intended to solve.

The `whl` file that is produced just contains the RDKit python files and the binary extension modules. It does not contain either the RDKit shared libraries or the boost shared libraries that the extension modules need. 

To demonstrate that, if I `pip install` the `whl` file after building it on my centos7 VM I get:
```
Python 3.6.6 (default, Mar 29 2019, 00:03:27) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-36)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import rdkit
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/src/rdkit/build_pip/rdk-env/lib64/python3.6/site-packages/rdkit/__init__.py", line 2, in <module>
    from .rdBase import rdkitVersion as __version__
ImportError: libRDKitRDBoost.so.1: cannot open shared object file: No such file or directory
```
This shows all the missing shared libraries:
```
(rdk-env) [root@e5780db95986 dist]# ldd /src/rdkit/build_pip/rdk-env/lib64/python3.6/site-packages/rdkit/rdBase.so 
	linux-vdso.so.1 =>  (0x00007fff243bf000)
	libRDKitRDBoost.so.1 => not found
	libpython3.6m.so.1.0 => /lib64/libpython3.6m.so.1.0 (0x00007f54b3121000)
	libboost_python36.so.1.69.0 => not found
	libboost_serialization.so.1.69.0 => not found
	libRDKitRDGeneral.so.1 => not found
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f54b2f05000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f54b2bfe000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f54b28fc000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f54b26e6000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f54b2319000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f54b2115000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007f54b1f12000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f54b3944000)
```
So, given that this `whl` isn't something that would be useful in pypi (because of the problems above),
as far as I can tell, the only thing this does is allow someone who has already done an RDKit build and installed the RDKit libraries somewhere on their LD_LIBRARY_PATH to do a `pip install` into their local python environment/virtual environment.

Do I have that right?

**Comment by UnixJunkie** [comment]
It would be amazing if the pip installer is really standalone.

**Comment by maksbotan** [comment]
@greglandrum thank you for looking into this!

You are right, resulting wheel does not contain RDKit libraries. Intended workflow is like this:

- Install rdkit into system with brew, apt-get or by building from source
- Create a virtualenv
- Install wheel there

This way you'll get python bindings in your virtualenv using rdkit libraries somewhere on the system (thanks to RPATH and stuff).

For example, some tutorials on getting opencv on macos (see https://medium.com/@nuwanprabhath/installing-opencv-in-macos-high-sierra-for-python-3-89c79f0a246a) recommend basically the same approach.

I'd like to emphasize that what we are trying to get with this is to **register rdkit in pip's package database**. That alone is a great improvement, since then rdkit's consumers can add rdkit to their `install_requires`, pin versions, etc.  We, for one, would benefit from this in our company a lot.

Now, about making standalone wheels. I'd certainly love to see this done, as it would be the simplest way for consumers to get rdkit. However, it looks impossible -- I've spent a long time trying to do that, in several different ways. The problem is that we either need to package boost into our wheel (which is plain wrong, let's not do that) or link boost statically (less wrong, but still _ewww_).

And it seems that boost-python is unsuited for static linking. Of course, I can take one more look in this if you wish.

@njzjz I hope this answers your question as well! I.e. it makes no sense to upload this on PyPI.

**Comment by greglandrum** [comment]
@maksbotan : I'm glad we agree with each other. :-)

Since I think I can understand where it's useful and since it doesn't seem to interfere with any of the other builds (though I still need to confirm that), I have no objections to merging this as long as there's a bit of documentation about how to use it. Could you please add a short section at the end of this: https://github.com/rdkit/rdkit/blob/master/Docs/Book/Install.md with some documentation about how to use the wheels?

**Comment by maksbotan** [comment]
Done! Let me know if I should change the wording.

**Comment by greglandrum** [comment]
> Done! Let me know if I should change the wording.

I made some suggestions in a PR against your branch. Feel free to correct those if I ended up saying anything incorrect.

**Comment by maksbotan** [comment]
@greglandrum well, I wanted to make it clear that

1. rdkit is registered with pip even with binary installs, i think this is important.
2. wheel is available on certain binary distributions, like homebrew (I plan to write that in brew's postinstall message)

Will you try to rewrite it or should I try again?

**Comment by greglandrum** [comment]
I'm not sure I understand point 1, so it's probably best if you try that part. :-)

**Comment by greglandrum** [comment]
hmm, I think I see where you want to go, but it's still confusing to me. I will think about it a bit more and try and come up with a suggestion tomorrow.

**Comment by maksbotan** [comment]
Great, thanks :)
I think you are able to write it more clear for RDKit users than me.

**Comment by maksbotan** [comment]
@greglandrum hello! how is it going? :)

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,7 @@ add_dependencies(rdkit_base catch)
 
 option(RDK_BUILD_SWIG_WRAPPERS "build the SWIG wrappers" OFF )
 option(RDK_BUILD_PYTHON_WRAPPERS "build the standard python wrappers" ON )
+option(RDK_INSTALL_PYTHON_PACKAGE "install Python wrappers via pip to the selected python" ON )
```
Comment (greglandrum) [inline CMakeLists.txt]: This should be `OFF`, since it's not likely to be part of the standard build process for people at the moment

Comment (maksbotan) [reply CMakeLists.txt]: Well, no, see what it does here:

https://github.com/rdkit/rdkit/pull/2690/files#diff-af3b638bc2a3e6c650974192a53c7291R300

It does exactly the same as old version did -- install Python bindings to `CMAKE_INSTALL_PREFIX/DESTDIR`.
Actually, I would argue that this option can be removed at all, making it the default behavior.

File: Docs/Book/Install.md [hunk 2]
```diff
@@ -499,27 +499,25 @@ This section assumes that python is installed in `C:\Python27`, that the boost l
...
+If you are building the RDKit yourself or using certain package managers, it is possible to install 
+the binaries using Python's package manager `pip`. By doing this pip becomes aware of the RDKit's
+existence, allowing you to use the RDKit with virtualenvs or add it to the `install_requires` lists for 
+other local packages' `setup.py` files.
 
-Internally, this is done by creating `*.whl` package and installing it at `make install` step with
-`pip` (when `-DRDK_INSTALL_PYTHON_PACKAGE` CMake option is `ON`, which is the default).
+Using the RDKit together with pip requires you to have a `.whl` file containing the RDKit's python files 
+and binary extension modules. If you do your own build, you can create this file by providing the 
+`-DRDK_INSTALL_PYTHON_PACKAGE=ON` argument to cmake (along with your usual other cmake arguments) and then
```
Comment (maksbotan) [inline Docs/Book/Install.md]: The wheel gets built unconditionally, this option only installs it to selected python's site-packages.

**Comment by psychemedia** [comment]
It would be so useful if I could just run:

`apt-get install -y python-rdkit librdkit1 rdkit-data`

and have:

`import rdkit` work rather than throw package not found errors. (What paths need setting? Where? Installing conda and then installing rdkit from conda feels like it should be unnecessary...)

**Comment by UnixJunkie** [comment]
@psychemedia On which Linux distribution are you?
If your distribution is Ubuntu or Debian but doesn't provide recent (i.e. supporting Python3) rdkit packages, you can try this:
```
git clone https://github.com/rdkit/rdkit.git
cd rdkit
./Scripts/create_deb_packages.sh
sudo dpkg -i build/*.deb
```
Of course, uninstall all rdkit packages on the system, if any, prior to this (because the packages will conflict).

**Comment by psychemedia** [comment]
@UnixJunkie It's debian, I think,  a third party circle.ci image. Will try your suggestion; thanks.

**Comment by UnixJunkie** [comment]
@psychemedia Also, asking to your distribution packagers to update the rdkit package they ship might not hurt (and even benefit more people).

**Comment by C-nit** [comment]
While installing the python bindings after building solves the issue of registering the package with pip, the different issue of installing rdkit via pip remains.
Having standalone wheels on pypi would make the installation and use of rdkit as dependency much easier for many, without the need to build.
The workflow here has been compared to an installation of opencv. Funnily enough there is [opencv-python](https://github.com/skvark/opencv-python) with over a thousand stars simply providing [wheels](https://pypi.org/project/opencv-python/#files) (lacking a source installation).

I appreciate this is not easily done and has been tried.
>Sadly, making self-contained rdkit wheel is rather challenging. As far as I can tell, you cannot link boost-python statically. I tried and I failed - I achieved linking without warnings, but it all broke in runtime.
> Another way is to bundle boost dynamic libraries in the wheel. This most certainly can be done. If you like this idea, I can do that. But it doesn't seem too nice for me.

opencv-python adds the shared libraries as well to do the job. It's using [delocate](https://pypi.org/project/delocate/) for linux and [auditwheel](https://pypi.org/project/auditwheel/) for MacOS.

**Comment by maksbotan** [comment]
Well I, as Gentoo developer, strongly dislike bundling anything. But, if rdkit community prefers this way, let's go opencv-python route. Let me study their scripts and adapt them to rdkit. Agreed?

But anyway, to do so this PR still has to be merged, because to bundle libs in the wheel, we must have one in the first place :)

**Comment by hsiaoyi0504** [comment]
Based on my previous experience in #1812, I totally agreed that we should build all things when `pip install rdkit` is invoked and I am thinking what rdkit community prefers and what @maksbotan mentions are not totally conflicting each other. We can make main `pip install rdkit` install all the stuff and behind the scenes, it install a set of python packages, might be called `rdkit-python`, `rdkit-boost` etc. For the users who prefer to install rdkit python binding library only, just install it through `pip install rdkit-python`.

**Comment by dmyersturnbull** [comment]
I love the idea @hsiaoyi0504 just proposed. Having a pip package that makes a fully sufficient, usable package is really important.

If you have dependencies on rdkit and on packages not available on Anaconda / Conda-Forge, it's quite an issue. My packages all used [Poetry](https://python-poetry.org/) and were trivial to install with just `pip`, but needing Conda for one package really throws a wrench into that. I can either have Conda overwrite my poetry-managed dependencies or the other way around, so there's no way to get a stable and reproducible build.

**Comment by UnixJunkie** [comment]
This should probably be merged. Some people want that pip can see that rdkit is installed, even if the installation was not managed by pip. They require that a .egg-info directory is present, once the rdkit python bindings have been installed.

**Comment by flying-sheep** [comment]
Definitely. Everything depending on RDkit is harder to test and use because there is no `.egg-info`/`.dist-info` directory.

**Comment by tbabej** [comment]
@UnixJunkie I think we should actually avoid merging this, creating portable binary wheels is possible, as evidence by the recent efforts on #1812 (also see [0] and [1]).

[0] https://github.com/kuelumbus/rdkit_platform_wheels
[1] https://github.com/ProteinQure/rdkit-wheels

**Comment by C-nit** [comment]
> Some people want that pip can see that rdkit is installed, even if the installation was not managed by pip.
While not really familiar with the PR, I fully agree with the quoted comment.

When developing a package it would be great to have in the setup.py `install_requires=["rdkit"]`. On installation this dependency has to be met.
- Either it's there already and I don't care whether it was with conda or pip or whatever. This is a python packaging issue.
- Or it's missing and pip will look it up on an index-server (pypi). Not finding it would error and that's a good thing to catch now.

Even better if there is something to be found on pypi to fulfil the dependency, be it a source dist or built wheels (like rdkit-pypi solved already) or both. This is a pip/pypi issue.

The only way we can check for rdkit being available on installation in python now is with `install_requires=["rdkit-pypi"]`. But with an existing conda installation this ends up with two installations in the same env.

@greglandrum
Fixing the python packaging issue should be a first, but as follow up providing as well something on pypi would be great to unify again the python installations depending on rdkit.

**Comment by hadim** [comment]
We also have the issue that any pip package relying on rdkit will override the conda installed rdkit package. This is an issue with datamol for example and for now we decided to not make rdkit a dep in the setup.py of datamol.

**Comment by greglandrum** [comment]
I am happy to review a PR which addresses the packaging problems people are having, but I don’t have the time to invest in learning the ins and outs of this stuff

**Comment by greglandrum** [comment]
I’m closing this specific PR because the problem of having a pip installable RDKit is resolved

**Comment by greglandrum** [comment]
Closing due to lack of activity. Feel free to reopen if anyone has the time and inclination to revive this one.

**Comment by hadim** [comment]
Hopefully, the packaging issue should be resolved by https://github.com/conda-forge/rdkit-feedstock/pull/143

**Pull request closed without merging**